File: tests/benchmarks/fixtures/sample.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"os"
	"sync"
	"time"
)

// Configuration constants
const (
	DefaultTimeout = 5000
	MaxRetries     = 3
	BatchSize      = 100
)

// logMsg is a simple logging wrapper for benchmark testing.
func logMsg(args ...interface{}) {
	fmt.Println(args...)
}

// Config holds application configuration.
type Config struct {
	Host           string `json:"host"`
	Port           int    `json:"port"`
	Debug          bool   `json:"debug"`
	MaxConnections int    `json:"max_connections"`
}

// User represents a user entity.
type User struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Email     string    `json:"email"`
	Role      string    `json:"role"`
	CreatedAt time.Time `json:"created_at"`
}

// QueryResult holds paginated query results.
type QueryResult struct {
	Data     []User `json:"data"`
	Total    int    `json:"total"`
	Page     int    `json:"page"`
	PageSize int    `json:"page_size"`
}

// EventHandler is a function type for handling events.
type EventHandler func(event string, data interface{})

// DatabaseConnection manages database connections.
type DatabaseConnection struct {
	config     Config
	connected  bool
	retryCount int
	mu         sync.Mutex
}

// NewDatabaseConnection creates a new database connection.
func NewDatabaseConnection(config Config) *DatabaseConnection {
	logMsg("DatabaseConnection created with config:", config)
	return &DatabaseConnection{config: config}
}

// Connect establishes the database connection.
func (db *DatabaseConnection) Connect() error {
	db.mu.Lock()
	defer db.mu.Unlock()
	fmt.Printf("Connecting to %s:%d...\n", db.config.Host, db.config.Port)
	db.connected = true
	db.retryCount = 0
	logMsg("Connected successfully")
	return nil
}

// Disconnect closes the database connection.
func (db *DatabaseConnection) Disconnect() error {
	db.mu.Lock()
	defer db.mu.Unlock()
	logMsg("Disconnecting...")
	db.connected = false
	logMsg("Disconnected")
	return nil
}

// IsConnected returns the connection status.
func (db *DatabaseConnection) IsConnected() bool {
	db.mu.Lock()
	defer db.mu.Unlock()
	return db.connected
}

// UserRepository provides user data access.
type UserRepository struct {
	db    *DatabaseConnection
	cache map[string]*User
	mu    sync.RWMutex
}

// NewUserRepository creates a new user repository.
func NewUserRepository(db *DatabaseConnection) *UserRepository {
	logMsg("UserRepository initialized")
	return &UserRepository{
		db:    db,
		cache: make(map[string]*User),
	}
}

// FindByID retrieves a user by ID.
func (r *UserRepository) FindByID(id string) (*User, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	fmt.Printf("Finding user by id: %s\n", id)
	user, ok := r.cache[id]
	if ok {
		fmt.Printf("Cache hit for user %s\n", id)
	} else {
		fmt.Printf("Cache miss for user %s, querying database\n", id)
	}
	return user, ok
}

// FindByEmail retrieves a user by email.
func (r *UserRepository) FindByEmail(email string) (*User, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	fmt.Printf("Finding user by email: %s\n", email)
	for _, user := range r.cache {
		if user.Email == email {
			return user, true
		}
	}
	return nil, false
}

// FindAll returns paginated users.
func (r *UserRepository) FindAll(page, pageSize int) QueryResult {
	r.mu.RLock()
	defer r.mu.RUnlock()
	fmt.Printf("Fetching users page %d, size %d\n", page, pageSize)
	users := make([]User, 0, len(r.cache))
	for _, u := range r.cache {
		users = append(users, *u)
	}
	start := (page - 1) * pageSize
	end := start + pageSize
	if end > len(users) {
		end = len(users)
	}
	if start > len(users) {
		start = len(users)
	}
	return QueryResult{
		Data:     users[start:end],
		Total:    len(users),
		Page:     page,
		PageSize: pageSize,
	}
}

// Create adds a new user.
func (r *UserRepository) Create(user *User) *User {
	r.mu.Lock()
	defer r.mu.Unlock()
	fmt.Printf("Creating user: %s\n", user.Name)
	r.cache[user.ID] = user
	fmt.Printf("User %s created successfully\n", user.Name)
	return user
}

// Update modifies an existing user.
func (r *UserRepository) Update(id string, name string, email string) (*User, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	fmt.Printf("Updating user %s\n", id)
	user, ok := r.cache[id]
	if !ok {
		fmt.Printf("User %s not found for update\n", id)
		return nil, false
	}
	if name != "" {
		user.Name = name
	}
	if email != "" {
		user.Email = email
	}
	fmt.Printf("User %s updated successfully\n", id)
	return user, true
}

// Delete removes a user.
func (r *UserRepository) Delete(id string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	fmt.Printf("Deleting user %s\n", id)
	_, ok := r.cache[id]
	if ok {
		delete(r.cache, id)
		fmt.Printf("User %s deleted\n", id)
	} else {
		fmt.Printf("User %s not found\n", id)
	}
	return ok
}

// GetCacheSize returns the number of cached users.
func (r *UserRepository) GetCacheSize() int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.cache)
}

// EventBus provides pub/sub event handling.
type EventBus struct {
	handlers map[string][]EventHandler
	eventLog []EventEntry
	mu       sync.RWMutex
}

// EventEntry records a dispatched event.
type EventEntry struct {
	Event     string    `json:"event"`
	Timestamp time.Time `json:"timestamp"`
}

// NewEventBus creates a new event bus.
func NewEventBus() *EventBus {
	logMsg("EventBus initialized")
	return &EventBus{
		handlers: make(map[string][]EventHandler),
	}
}

// Register adds a handler for an event.
func (eb *EventBus) Register(event string, handler EventHandler) {
	eb.mu.Lock()
	defer eb.mu.Unlock()
	fmt.Printf("Registering handler for event: %s\n", event)
	eb.handlers[event] = append(eb.handlers[event], handler)
}

// Dispatch triggers all handlers for an event.
func (eb *EventBus) Dispatch(event string, data interface{}) {
	eb.mu.RLock()
	defer eb.mu.RUnlock()
	fmt.Printf("Dispatching event: %s\n", event)
	eb.eventLog = append(eb.eventLog, EventEntry{Event: event, Timestamp: time.Now()})
	for _, handler := range eb.handlers[event] {
		handler(event, data)
	}
}

// ValidateEmail checks if an email address is valid.
func ValidateEmail(email string) bool {
	fmt.Printf("Validating email: %s\n", email)
	return len(email) > 3 && containsChar(email, '@') && containsChar(email, '.')
}

func containsChar(s string, c byte) bool {
	for i := 0; i < len(s); i++ {
		if s[i] == c {
			return true
		}
	}
	return false
}

// GenerateID creates a random identifier.
func GenerateID() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, 15)
	for i := range b {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return string(b)
}

// FormatUser returns a formatted string for a user.
func FormatUser(user *User) string {
	fmt.Printf("Formatting user: %s\n", user.Name)
	return fmt.Sprintf("%s <%s> (%s)", user.Name, user.Email, user.Role)
}

// LoadConfig reads configuration from a JSON file.
func LoadConfig(path string) (*Config, error) {
	fmt.Printf("Loading config from %s\n", path)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("cannot read config: %w", err)
	}
	var config Config
	if err := json.Unmarshal(data, &config); err != nil {
		return nil, fmt.Errorf("cannot parse config: %w", err)
	}
	logMsg("Config loaded:", config)
	return &config, nil
}

// SaveConfig writes configuration to a JSON file.
func SaveConfig(path string, config *Config) error {
	logMsg("Saving config to", path)
	data, err := json.MarshalIndent(config, "", "  ")
	if err != nil {
		return fmt.Errorf("cannot marshal config: %w", err)
	}
	if err := os.WriteFile(path, data, 0644); err != nil {
		return fmt.Errorf("cannot write config: %w", err)
	}
	logMsg("Config saved")
	return nil
}

// ProcessItems splits items into batches.
func ProcessItems(items []interface{}, batchSize int) [][]interface{} {
	fmt.Printf("Processing %d items in batches of %d\n", len(items), batchSize)
	var batches [][]interface{}
	for i := 0; i < len(items); i += batchSize {
		end := i + batchSize
		if end > len(items) {
			end = len(items)
		}
		batches = append(batches, items[i:end])
	}
	fmt.Printf("Created %d batches\n", len(batches))
	return batches
}

func main() {
	log.SetFlags(log.LstdFlags | log.Lshortfile)
	config := Config{
		Host:           "localhost",
		Port:           5432,
		Debug:          true,
		MaxConnections: 10,
	}
	db := NewDatabaseConnection(config)
	if err := db.Connect(); err != nil {
		log.Fatal("Failed to connect:", err)
	}
	defer db.Disconnect()

	repo := NewUserRepository(db)
	bus := NewEventBus()

	user := &User{
		ID:        GenerateID(),
		Name:      "Alice",
		Email:     "alice@example.com",
		Role:      "admin",
		CreatedAt: time.Now(),
	}
	repo.Create(user)
	bus.Dispatch("user.created", user)
	fmt.Println(FormatUser(user))
}

File: tests/benchmarks/fixtures/sample.rs
use std::collections::HashMap;
use std::fmt;
use std::fs;
use std::io::{self, Read, Write};
use std::sync::{Arc, Mutex, RwLock};
use std::time::{Duration, Instant};

// Configuration constants
const DEFAULT_TIMEOUT: u64 = 5000;
const MAX_RETRIES: u32 = 3;
const BATCH_SIZE: usize = 100;

#[derive(Debug, Clone)]
struct Config {
    host: String,
    port: u16,
    debug: bool,
    max_connections: u32,
}

#[derive(Debug, Clone)]
struct User {
    id: String,
    name: String,
    email: String,
    role: String,
    created_at: Instant,
}

#[derive(Debug)]
struct QueryResult {
    data: Vec<User>,
    total: usize,
    page: usize,
    page_size: usize,
}

type EventHandler = Box<dyn Fn(&str, &dyn fmt::Debug) + Send + Sync>;

struct DatabaseConnection {
    config: Config,
    connected: bool,
    retry_count: u32,
}

impl DatabaseConnection {
    fn new(config: Config) -> Self {
        println!("DatabaseConnection created with config: {:?}", config);
        Self {
            config,
            connected: false,
            retry_count: 0,
        }
    }

    fn connect(&mut self) -> Result<(), String> {
        println!("Connecting to {}:{}...", self.config.host, self.config.port);
        self.connected = true;
        self.retry_count = 0;
        println!("Connected successfully");
        Ok(())
    }

    fn disconnect(&mut self) {
        println!("Disconnecting...");
        self.connected = false;
        println!("Disconnected");
    }

    fn is_connected(&self) -> bool {
        self.connected
    }
}

struct UserRepository {
    db: Arc<Mutex<DatabaseConnection>>,
    cache: RwLock<HashMap<String, User>>,
}

impl UserRepository {
    fn new(db: Arc<Mutex<DatabaseConnection>>) -> Self {
        println!("UserRepository initialized");
        Self {
            db,
            cache: RwLock::new(HashMap::new()),
        }
    }

    fn find_by_id(&self, id: &str) -> Option<User> {
        println!("Finding user by id: {}", id);
        let cache = self.cache.read().unwrap();
        let result = cache.get(id).cloned();
        if result.is_some() {
            println!("Cache hit for user {}", id);
        } else {
            println!("Cache miss for user {}, querying database", id);
        }
        result
    }

    fn find_by_email(&self, email: &str) -> Option<User> {
        println!("Finding user by email: {}", email);
        let cache = self.cache.read().unwrap();
        cache.values().find(|u| u.email == email).cloned()
    }

    fn find_all(&self, page: usize, page_size: usize) -> QueryResult {
        println!("Fetching users page {}, size {}", page, page_size);
        let cache = self.cache.read().unwrap();
        let users: Vec<User> = cache.values().cloned().collect();
        let start = (page - 1) * page_size;
        let end = (start + page_size).min(users.len());
        let start = start.min(users.len());
        QueryResult {
            data: users[start..end].to_vec(),
            total: users.len(),
            page,
            page_size,
        }
    }

    fn create(&self, user: User) -> User {
        println!("Creating user: {}", user.name);
        let mut cache = self.cache.write().unwrap();
        cache.insert(user.id.clone(), user.clone());
        println!("User {} created successfully", user.name);
        user
    }

    fn update(&self, id: &str, name: Option<&str>, email: Option<&str>) -> Option<User> {
        println!("Updating user {}", id);
        let mut cache = self.cache.write().unwrap();
        let user = cache.get_mut(id);
        match user {
            Some(u) => {
                if let Some(n) = name {
                    u.name = n.to_string();
                }
                if let Some(e) = email {
                    u.email = e.to_string();
                }
                println!("User {} updated successfully", id);
                Some(u.clone())
            }
            None => {
                println!("User {} not found for update", id);
                None
            }
        }
    }

    fn delete(&self, id: &str) -> bool {
        println!("Deleting user {}", id);
        let mut cache = self.cache.write().unwrap();
        let removed = cache.remove(id).is_some();
        if removed {
            println!("User {} deleted", id);
        } else {
            println!("User {} not found", id);
        }
        removed
    }

    fn get_cache_size(&self) -> usize {
        self.cache.read().unwrap().len()
    }
}

struct EventBus {
    handlers: HashMap<String, Vec<EventHandler>>,
    event_log: Vec<(String, Instant)>,
}

impl EventBus {
    fn new() -> Self {
        println!("EventBus initialized");
        Self {
            handlers: HashMap::new(),
            event_log: Vec::new(),
        }
    }

    fn register(&mut self, event: &str, handler: EventHandler) {
        println!("Registering handler for event: {}", event);
        self.handlers
            .entry(event.to_string())
            .or_default()
            .push(handler);
    }

    fn dispatch(&mut self, event: &str, data: &dyn fmt::Debug) {
        println!("Dispatching event: {}", event);
        self.event_log.push((event.to_string(), Instant::now()));
        if let Some(handlers) = self.handlers.get(event) {
            for handler in handlers {
                handler(event, data);
            }
        }
    }
}

fn validate_email(email: &str) -> bool {
    println!("Validating email: {}", email);
    email.contains('@') && email.contains('.') && email.len() > 3
}

fn generate_id() -> String {
    use std::time::SystemTime;
    let duration = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap();
    format!("{:x}", duration.as_nanos())[..15].to_string()
}

fn format_user(user: &User) -> String {
    println!("Formatting user: {}", user.name);
    format!("{} <{}> ({})", user.name, user.email, user.role)
}

fn load_config(path: &str) -> Result<Config, String> {
    println!("Loading config from {}", path);
    let content = fs::read_to_string(path).map_err(|e| format!("cannot read config: {}", e))?;
    let parsed: serde_json::Value =
        serde_json::from_str(&content).map_err(|e| format!("cannot parse config: {}", e))?;
    let config = Config {
        host: parsed["host"].as_str().unwrap_or("localhost").to_string(),
        port: parsed["port"].as_u64().unwrap_or(8080) as u16,
        debug: parsed["debug"].as_bool().unwrap_or(false),
        max_connections: parsed["max_connections"].as_u64().unwrap_or(10) as u32,
    };
    println!("Config loaded: {:?}", config);
    Ok(config)
}

fn save_config(path: &str, config: &Config) -> Result<(), String> {
    println!("Saving config to {}", path);
    let json = format!(
        r#"{{"host":"{}","port":{},"debug":{},"max_connections":{}}}"#,
        config.host, config.port, config.debug, config.max_connections
    );
    fs::write(path, json).map_err(|e| format!("cannot write config: {}", e))?;
    println!("Config saved");
    Ok(())
}

fn process_items<T: Clone>(items: &[T], batch_size: usize) -> Vec<Vec<T>> {
    println!(
        "Processing {} items in batches of {}",
        items.len(),
        batch_size
    );
    let batches: Vec<Vec<T>> = items.chunks(batch_size).map(|c| c.to_vec()).collect();
    println!("Created {} batches", batches.len());
    batches
}

fn retry_operation<F, T, E>(operation: F, max_retries: u32, delay_ms: u64) -> Result<T, E>
where
    F: Fn() -> Result<T, E>,
    E: fmt::Display,
{
    println!("Starting operation with {} retries", max_retries);
    for attempt in 1..=max_retries {
        match operation() {
            Ok(result) => {
                println!("Operation succeeded on attempt {}", attempt);
                return Ok(result);
            }
            Err(e) => {
                println!("Attempt {} failed: {}", attempt, e);
                if attempt == max_retries {
                    return Err(e);
                }
                std::thread::sleep(Duration::from_millis(delay_ms * attempt as u64));
            }
        }
    }
    unreachable!()
}

fn main() {
    let config = Config {
        host: "localhost".to_string(),
        port: 5432,
        debug: true,
        max_connections: 10,
    };

    let db = Arc::new(Mutex::new(DatabaseConnection::new(config)));
    db.lock().unwrap().connect().unwrap();

    let repo = UserRepository::new(db.clone());
    let mut bus = EventBus::new();

    let user = User {
        id: generate_id(),
        name: "Alice".to_string(),
        email: "alice@example.com".to_string(),
        role: "admin".to_string(),
        created_at: Instant::now(),
    };

    repo.create(user.clone());
    bus.dispatch("user.created", &user.name);
    println!("{}", format_user(&user));
}
